gemd/ingest/impl.py

```diff
@@ -0,0 +1,208 @@
+    EmpiricalFormula, Smiles
+from gemd.entity.value.base_value import BaseValue
+from typing import Union
+
+
+def make_node(name: str,
+              *,
+              process_name: str = None,
+              process_template: ProcessTemplate = None,
+              material_template: MaterialTemplate = None):
+    """
+    Generate a material-process spec-run quadruple.
+
+    Parameters
+    ----------
+    name: str
+        Name of the MaterialRun and MaterialSpec.
+    process_name: str
+        Name of the ProcessRun and ProcessSpec.  Defaults to
+        `process_template.name` if `process_template` is defined, else `name`.
+    process_template: ProcessTemplate
+        ProcessTemplate for the quadruple.
+    material_template: MaterialTemplate
+        MaterialTemplate for the quadruple.
+
+    Returns
+    --------
+    MaterialRun
+        A MaterialRun with linked processes, specs and templates
+
+    """
+    if process_name is None:
+        if process_template is None:
+            process_name = name
+        else:
+            process_name = process_template.name
+
+    my_process_spec = ProcessSpec(
+        name=process_name,
+        template=process_template
+    )
+
+    my_process_run = ProcessRun(
+        name=process_name,
+        spec=my_process_spec
+    )
+
+    my_mat_spec = MaterialSpec(
+        name=name,
+        process=my_process_spec,
+        template=material_template
+    )
+
+    my_mat_run = MaterialRun(
+        name=name,
+        process=my_process_run,
+        spec=my_mat_spec
+    )
+    return my_mat_run
+
+
+def add_edge(input_material: MaterialRun,
+             output_material: MaterialRun,
+             *,
+             name: str = None):
+    """
+    Connect two material-process spec-run quadruples with ingredients.
+
+    Parameters
+    ----------
+    input_material: MaterialRun
+        The `material` for the returned IngredientRun
+    output_material: MaterialRun
+        The `process` for the returned IngredientRun will be
+        `output_material.process`
+    name: str
+        The ingredient name.  Defaults to `input_material.name`.
+
+    Returns
+    --------
+    MaterialRun
+        A MaterialRun with linked processes, specs and templates
+
+    """
+    if name is None:
+        name = input_material.name
+    my_ingredient_spec = IngredientSpec(name=name,
+                                        process=output_material.spec.process,
+                                        material=input_material.spec
+                                        )
+    my_ingredient_spec = IngredientRun(spec=my_ingredient_spec,
+                                       process=output_material.process,
+                                       material=input_material
+                                       )
+    return my_ingredient_spec
+
+
+def add_measurement(material: MaterialRun,
+                    name: str = None,
```

Comment:
Shouldn't this argument come after the asterisk?

Reply:
I was thinking it made sense to support the optional second argument -- the most common use case felt like `add_measurement(node, 'Density')` -- but if that doesn't feed like the correct flow to you, moving the argument is reasonable.  It's always easier to add a positional argument than remove one.

Reply:
It's not obvious to me whether it would be more common to specify the measurement name via a string or via a template.

Reply:
Changed per request.